File: app/src/main/java/ca/gainzassist/models/db/WorkoutDatabase.java
package ca.gainzassist.models.db;

import androidx.sqlite.db.SupportSQLiteDatabase;
import androidx.room.*;
import android.content.Context;
import androidx.annotation.NonNull;
import ca.gainzassist.models.Exercise;
import ca.gainzassist.models.Session;
import ca.gainzassist.models.ExerciseSet;
import ca.gainzassist.models.Workout;

@Database(entities = {Workout.class, Exercise.class, ExerciseSet.class, Session.class},
          version = 1,
          exportSchema = false)
public abstract class WorkoutDatabase extends RoomDatabase {

    public abstract WorkoutDao workoutDao();
    public abstract ExerciseDao exerciseDao();
    public abstract SetDao setDao();
    public abstract SessionDao sessionDao();

    private static volatile WorkoutDatabase INSTANCE;

    private static RoomDatabase.Callback sRoomDatabaseCallback =
            new RoomDatabase.Callback(){

                @Override
                public void onOpen (@NonNull SupportSQLiteDatabase db){
                    super.onOpen(db);
                }
            };

    public static WorkoutDatabase getDatabase(final Context context) {
        if (INSTANCE == null) {
            synchronized (WorkoutDatabase.class) {
                if (INSTANCE == null) {
                    INSTANCE = Room.databaseBuilder(context.getApplicationContext(),
                            WorkoutDatabase.class, "workout_database")
                            .addCallback(sRoomDatabaseCallback)
                            .build();
                }
            }
        }
        return INSTANCE;
    }
}

File: app/src/main/java/ca/gainzassist/models/db/WorkoutDatabase.kt
package ca.gainzassist.models.db

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import ca.gainzassist.models.Exercise
import ca.gainzassist.models.ExerciseSet
import ca.gainzassist.models.Session
import ca.gainzassist.models.Workout

@Database(
    entities = [Workout::class, Exercise::class, ExerciseSet::class, Session::class],
    version = 1,
    exportSchema = false
)
abstract class WorkoutDatabase : RoomDatabase() {

    abstract fun workoutDao(): WorkoutDao
    abstract fun exerciseDao(): ExerciseDao
    abstract fun setDao(): SetDao
    abstract fun sessionDao(): SessionDao

    companion object {
        private const val DATABASE_NAME = "workout_database"

        @Volatile
        private var INSTANCE: WorkoutDatabase? = null

        private val sRoomDatabaseCallback = object : Callback() {
            /* no-op */
        }

        @JvmStatic
        fun getDatabase(context: Context): WorkoutDatabase {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: getDatabaseBuilder(context)
                    .addCallback(sRoomDatabaseCallback)
                    .build()
                    .also { INSTANCE = it }
            }
        }

        private fun getDatabaseBuilder(context: Context) = Room.databaseBuilder(
            context.applicationContext,
            WorkoutDatabase::class.java,
            DATABASE_NAME
        )
    }
}
